src/storage/src/source/postgres.rs

```diff
@@ -129,8 +129,38 @@ impl SourceRender for PostgresSourceConnection {
         Stream<G, (usize, HealthStatusUpdate)>,
         Rc<dyn Any>,
     ) {
-        let resume_upper =
-            Antichain::from_iter(config.source_resume_upper.iter().map(MzOffset::decode_row));
+        // Determined which collections need to be snapshot and which already have been.
+        let (mut to_snapshot, already_snapshot): (BTreeMap<_, _>, BTreeMap<_, _>) = config
+            .source_resume_upper
+            .iter()
+            .map(|(id, upper)| {
+                (
+                    *id,
+                    Antichain::from_iter(upper.iter().map(MzOffset::decode_row)),
+                )
+            })
+            .partition(|(_id, upper)| **upper == [MzOffset::minimum()]);
+
+        // We never snapshot the primary source, but its upper can influence the resume upper.
+        to_snapshot.remove(&config.id);
+
+        assert!(
+            to_snapshot.is_empty() ^ already_snapshot.is_empty(),
+            "until we support adding subsources only one of these sets may have elements"
```

Comment:
For this first point, we discussed my confusion between resumption frontiers and the flow of resume uppers we receive; I thought that the meet _and_ join of the subsources was the same--i.e. we only minted a timestamp binding when _all_ subsources committed, rather than what actually happens, which is minting a new timestamp when the first subsource committed. Other subsources are then guaranteed to use the same timestamp because their snapshots will be understood to complete at the same LSN as the first.

For the second point, moved those bits around.